c7n/resources/s3.py

```diff
@@ -375,6 +377,244 @@ def get_permissions(self):
+                                    (multipart_days_in_rule != multipart_days_in_policy):
+                                bad_bucket_lifecycle = True
+                        elif multipart_days_in_policy:
+                            bad_bucket_lifecycle = True
+
+                if not bad_bucket_lifecycle:
+                    results.append(b)
+
+        if self.cache_update is True:
+            shared_buckets = {b['Name']: b for b in shared_buckets}
+            self.manager._cache.save(CACHE_NAME, shared_buckets)
+
+        self.log.info("Found %d buckets with the lifecycle specified, out of %d buckets",
+                      len(results),
+                      len(buckets))
+
+        return results
+
+    def call_api(self, bucket):
+        bname = bucket['Name']
+        session = local_session(self.manager.session_factory)
+        s3 = bucket_client(session, bucket)
+
+        try:
+            shared_buckets = self.manager._cache.get(CACHE_NAME)
+            if shared_buckets:
+                if (bname in shared_buckets) and ('LifecyclePolicy' in shared_buckets[bname]):
+                    return bucket
+                else:
+                    lifecycle = s3.get_bucket_lifecycle_configuration(Bucket=bname)
+                    shared_buckets[bname] = bucket
+                    shared_buckets[bname]['LifecyclePolicy'] = lifecycle
+                    self.cache_update = True
+                    return shared_buckets[bname]
+            else:
+                lifecycle = s3.get_bucket_lifecycle_configuration(Bucket=bname)
+                bucket['LifecyclePolicy'] = lifecycle
+                self.cache_update = True
+                return bucket
+        except ClientError as e:
+            # Do nothing
+            if e.response['Error']['Code'] == 'NoSuchLifecycleConfiguration':
+                return
+            if e.response['Error']['Code'] == 'NoSuchBucket':
+                self.log.warn("Bucket `%s` no longer exists", bname)
+                return
+            raise
+
+
+def get_bucket_region(b):
+    bucket_region = b['Location']['LocationConstraint']
+    if bucket_region is None:
+        return 'us-east-1'
+    # EU = eu-west-1.
+    # From docs: https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGETlocation.html
+    elif bucket_region == "EU":
+        return 'eu-west-1'
+    else:
+        return bucket_region
+
+
+class S3BucketLifecycle(object):
+    """Object representing an S3 lifecycle rule."""
+
+    def __init__(self, prefix, rule_id):
+        self.lifecycle = {
+            'ID': rule_id,
+            'Status': 'Enabled',
+            'Filter': {
+                'Prefix': ''
+            }
+        }
+
+    def set_deletes_objects_days(self, days):
+        self.lifecycle['Expiration'] = {'Days': days}
+
+    def set_expired_delete_marker(self, value):
+        self.lifecycle['Expiration'] = {'ExpiredObjectDeleteMarker': value}
```

Comment:
Is there any chance that both `set_deletes_objects_days` and `set_expired_delete_marker` can be called?  If so, the second one called would overwrite the other.  (That may be on purpose, I don't fully understand the lifecycle policies)